### PR TITLE
[expo-cli][xdl] Check android.package before starting build

### LINKED ISSUE #256: `expo build:android` should check configuration more strictly before actual building process
## Description

screenshot 2018-12-13 at 11 51 52

## Expected Behavior

Information about missing configuration in project/`app.json` should be provided at the beginning.
`expo build:ios` is prompting about missing configuration in `app.json`.

## Observed Behavior

Project is being checked before building process and despite the fact configuration is not meeting requirements we're prompted with `everting is okay - proceeding`.

## Reproducible Demo

* `expo init`
* navigate to project and `yarn`
* `expo build:android`

### PR DESCRIPTION
**Pull request by jakubste**
Closes #256 

zrzut ekranu 2018-12-19 o 11 40 19

### CODE REVIEW

File: packages/expo-cli/src/commands/build/AndroidBuilder.js [hunk 1]
```diff
@@ -212,11 +212,20 @@ export default class AndroidBuilder extends BaseBuilder {
   async validateProject(options) {
     const publicUrl = options.publicUrl;
     const {
-      args: { sdkVersion },
+      args: { sdkVersion, androidPackage },
     } = publicUrl
-      ? await Exp.getThirdPartyInfoAsync(publicUrl)
-      : await Exp.getPublishInfoAsync(this.projectDir);
+      ? await Exp.getThirdPartyInfoAsync(publicUrl, 'android')
+      : await Exp.getPublishInfoAsync(this.projectDir, 'android');
```
Comment (fson) [inline packages/expo-cli/src/commands/build/AndroidBuilder.js]: `Exp.getThirdPartyInfoAsync` also needs to be updated to return the `androidPackage` field.

File: packages/xdl/src/Exp.js [hunk 2]
```diff
@@ -284,16 +284,20 @@ export async function getPublishInfoAsync(root: string): Promise<PublishInfo> {
   const remotePackageName = name;
   const remoteUsername = username;
   const remoteFullPackageName = `@${remoteUsername}/${remotePackageName}`;
-  const bundleIdentifierIOS = exp.ios ? exp.ios.bundleIdentifier : null;
+
+  const platformSpecific =
+    platform === 'android'
+      ? { androidPackage: _.get(exp, 'android.package', null) }
+      : { bundleIdentifierIOS: _.get(exp, 'ios.bundleIdentifier', null) };
```
Comment (fson) [inline packages/xdl/src/Exp.js]: To keep things simple, I suggest we just always return both of these fields (or `null` if they're not defined in the manifest) and remove the optional `platform` argument.

File: packages/expo-cli/src/commands/build/AndroidBuilder.js [hunk 3]
```diff
@@ -212,11 +212,20 @@ export default class AndroidBuilder extends BaseBuilder {
...
     await this.checkIfSdkIsSupported(sdkVersion, 'android');
+
+    if (!androidPackage) {
+      throw new XDLError(
+        ErrorCode.INVALID_OPTIONS,
+        `Your project must have an Android package set in app.json.
```
Comment (fson) [inline packages/expo-cli/src/commands/build/AndroidBuilder.js]: The error message should mention the field is `android.package`, or even better, prompt the user for the package name and write it to `app.json`, example: https://github.com/expo/expo-cli/blob/371be04175a676ae789544836ada6421a5d95937/packages/xdl/src/detach/Detach.js#L213-L225

Comment (wkozyra95) [reply packages/expo-cli/src/commands/build/AndroidBuilder.js]: updating a local android package would be confusing for users using self-hosted apps. We would update local app.json and deploy remote one.

**Comment by wkozyra95** [comment]
done in #367

**Pull request closed without merging**